README.md
# PyFABDB
[![Test](https://github.com/GonzalezAndrew/pyfabdb/actions/workflows/test.yml/badge.svg?branch=master)](https://github.com/GonzalezAndrew/pyfabdb/actions/workflows/test.yml)
[![pre-commit](https://github.com/GonzalezAndrew/pyfabdb/actions/workflows/pre-commit.yml/badge.svg?branch=master)](https://github.com/GonzalezAndrew/pyfabdb/actions/workflows/pre-commit.yml)

PyFABDB is a Python library to access the [Flesh and Blood DB API](https://fabdb.net/resources/api). The Python library enables you to access all public Flesh and Blood DB API endpoints using Python.

## Install
```
$ pip install pyfabdb
```

## Examples

```python
>>> from pyfabdb import pyfabdb
>>> fabdb = pyfabdb.PyFabdb()
>>> card = fabdb.get_card(id="absorb-in-aether-red")
>>> print(card["name"])
Absorb in Aether
```

## Contributing
All help is welcomed!

### 1. Setting up an environment
The easiest way to set up a test environment is to run:

1. `tox --devenv venv`
2. `source venv/bin/activate`

### 2. Running all the tests
After making your changes and creating new test for any changes, you will need to ensure all test pass. Running all the test can be done by running `tox -e py`.

### 3. Install pre-commit
With the environment created by tox, run:

1. `pip install pre-commit`
2. `pre-commit install`

### 4. Create a pull requests
Create a pull request following the pull request template. A maintainer will review the changes made, update the changelog and release the new version of pyfabdb.


CHANGELOG.md
v0.1.0 - 2021-09-02
===================
- initial release


tests/card_test.py
import pytest

from pyfabdb import pyfabdb


@pytest.fixture()
def fabdb():
    return pyfabdb.PyFabdb()


def test_get_card(fabdb):
    """ Test the get_card function """
    card = fabdb.get_card(id='absorb-in-aether-red')
    assert isinstance(card, dict)
    assert card['name'] == 'Absorb in Aether'
    assert card['type'] == 'defense'
    assert card['class'] == 'wizard'


def test_cards(fabdb):
    """ Test the cards function """
    cards = fabdb.cards()
    assert isinstance(cards, dict)
    assert len(cards['data']) > 0


def test_cards_by_pitch(fabdb):
    """ Test the cards function, param pitch """
    pitch_1 = fabdb.cards(pitch='1')
    assert isinstance(pitch_1, dict)
    assert len(pitch_1) > 0
    for card in pitch_1['data']:
        pitch = card['stats']['resource']
        assert pitch == '1' or pitch == 1

    pitch_2 = fabdb.cards(pitch='2')
    assert isinstance(pitch_2, dict)
    assert len(pitch_2) > 0
    for card in pitch_2['data']:
        pitch = card['stats']['resource']
        assert pitch == '2' or pitch == 2

    pitch_3 = fabdb.cards(pitch='3')
    assert isinstance(pitch_3, dict)
    assert len(pitch_3) > 0
    for card in pitch_3['data']:
        pitch = card['stats']['resource']
        assert pitch == '3' or pitch == 3

    # pitch can only be 1, 2, or 3
    pitch_4 = fabdb.cards(pitch='4')
    assert len(pitch_4['data']) == 0


def test_cards_by_keywords(fabdb):
    """ Test the cards function, param keywords"""
    keyword_1 = fabdb.cards(keywords='equipment')
    assert 'equipment' in keyword_1['data'][0]['keywords']


def test_cards_by_cost(fabdb):
    """ Test the cards function, param cost """
    cost_1 = fabdb.cards(cost='1')
    for card in cost_1['data']:
        cost = card['stats']['cost']
        assert cost == '1' or cost == 1

    cost_2 = fabdb.cards(cost='2')
    for card in cost_2['data']:
        cost = card['stats']['cost']
        assert cost == '2' or cost == 2

    cost_3 = fabdb.cards(cost='3')
    for card in cost_3['data']:
        cost = card['stats']['cost']
        assert cost == '3' or cost == 3

    # currently there is a bug with filtering for card with cost of 4
    # cost_4 = fabdb.cards(cost="4")
    # for card in cost_4["data"]:
    #     cost = card["stats"]["cost"]
    #     assert cost == "4" or cost == 4


def test_cards_by_rarity(fabdb):
    """ Test the cards function, param rarity """
    rarity_1 = fabdb.cards(rarity='C')
    # bug with rarity C, returns other rarity cards
    # for card in rarity_1["data"]:
    #     assert card["rarity"] == "C"
    assert len(rarity_1['data']) > 0

    rarity_2 = fabdb.cards(rarity='R')
    # bug with rarity R, returns other rarity cards
    # for card in rarity_2["data"]:
    #     assert card["rarity"] == "R"
    assert len(rarity_2['data']) > 0

    rarity_3 = fabdb.cards(rarity='S')
    for card in rarity_3['data']:
        assert card['rarity'] == 'S'
    assert len(rarity_3['data']) > 0

    rarity_4 = fabdb.cards(rarity='T')
    for card in rarity_4['data']:
        assert card['rarity'] == 'T'
    assert len(rarity_4['data']) > 0

    rarity_5 = fabdb.cards(rarity='L')
    for card in rarity_5['data']:
        assert card['rarity'] == 'L'
    assert len(rarity_5['data']) > 0

    rarity_6 = fabdb.cards(rarity='F')
    for card in rarity_6['data']:
        assert card['rarity'] == 'F'
    assert len(rarity_6['data']) > 0

    rarity_7 = fabdb.cards(rarity='P')
    # bug with rarity P, returns other rarity cards
    # for card in rarity_7["data"]:
    #     print(card["rarity"])
    # assert card["rarity"] == "P"
    assert len(rarity_7['data']) > 0


def test_cards_by_class(fabdb):
    """ Test the cards function, param card_class """
    brute = fabdb.cards(card_class='brute')
    assert len(brute['data']) > 0
    for cards in brute['data']:
        assert 'brute' in cards['keywords']

    guardian = fabdb.cards(card_class='guardian')
    assert len(guardian['data']) > 0
    for cards in guardian['data']:
        assert 'guardian' in cards['keywords']

    mechanologist = fabdb.cards(card_class='mechanologist')
    assert len(mechanologist['data']) > 0
    for cards in mechanologist['data']:
        assert 'mechanologist' in cards['keywords']

    ninja = fabdb.cards(card_class='ninja')
    assert len(ninja['data']) > 0
    for cards in ninja['data']:
        assert 'ninja' in cards['keywords']

    ranger = fabdb.cards(card_class='ranger')
    assert len(ranger['data']) > 0
    for cards in ranger['data']:
        assert 'ranger' in cards['keywords']

    runeblade = fabdb.cards(card_class='runeblade')
    assert len(runeblade['data']) > 0
    for cards in runeblade['data']:
        assert 'runeblade' in cards['keywords']

    warrior = fabdb.cards(card_class='warrior')
    assert len(warrior['data']) > 0
    for cards in warrior['data']:
        assert 'warrior' in cards['keywords']

    wizard = fabdb.cards(card_class='wizard')
    assert len(wizard['data']) > 0
    for cards in wizard['data']:
        assert 'wizard' in cards['keywords']


def test_cards_by_set(fabdb):
    """ Test the cards function, param set """
    wtr = fabdb.cards(set='WTR')
    assert len(wtr['data']) > 0

    arc = fabdb.cards(set='ARC')
    assert len(arc['data']) > 0

    cru = fabdb.cards(set='CRU')
    assert len(cru['data']) > 0

    mon = fabdb.cards(set='MON')
    assert len(mon['data']) > 0


tests/util_test.py
from pyfabdb.util import format_url


def test_format_url():
    """ Test the format_url utility function"""
    base_url = 'https://api.fabdb.net'
    path = '/cards'
    expected = base_url + path
    assert format_url(base_url, path) == expected


tests/cards_test.py
from pyfab import pyfab

fab = pyfab.PyFab()


def test_cards():
    """Test the cards endpoint"""
    all_cards = fab.cards()
    assert isinstance(all_cards, dict)


def test_decks():
    """Test the decks endpoint"""
    decks = fab.decks(slug='qYyXnByL')
    assert isinstance(decks, dict)


test_cards()
test_decks()


pyfab/pyfab.py
from pyfab.client import Client


class PyFab(Client):
    def __init__(self, url=None, heade=None):
        super().__init__(url='https://api.fabdb.net', header={'Accept': 'application/json'})

    def cards(self) -> dict:
        ''' Get cards from fab API.
        :param page: When paginating through data sets, specifies the page.
            Integer.
        :param per_page: Specify the number of records per result set.
            Integer. Max-100
        :param keywords: A string representing a search term:
            String
        :params pitch: Search for a card with a given pitch.
            String. Valid values are: (1, 2, or 3)
        :params cost: Search for a card with a given cost.
            String. Valid values are: (1, 2, 3, or 4+)
        :params class: Search for a card that matches the required hero class.
           String. Valid values are: brute, guardian, mechanologist, ninja, ranger, runeblade, warrior, wizard.
        :params rarity: Search for a card that matches the specified rarity.
            String. Valid values are: C, R, S, T, L, F, P
        :params set: Search for cards from a given set.
            String. Valid values are: WTR, ARC, CRU, MON
        :return: A list of dictionaries with the cards' data.
        :rtype: list[dict{}]
            Example Returned Data:
            {
                "current_page":19,
                "data": [
                    {
                        "identifier":"WTR219",
                        "name":"Nimblism",
                        "keywords":["generic","action"],
                        "stats":{"cost":"0","defense":"2","resource":"2"},
                    },
                    ...
                ],
                "first_page_url":"/cards?per_page=25&page=1",
                "from":451,
                "next_page_url":null,
                "path":"/cards",
                "per_page":"25",
                "prev_page_url":"/cards?per_page=25&page=18",
                "to":457
            }
        '''
        return self.get(path='/cards')

    def get_card(self, id: str = 'absorb-in-aether-red') -> dict:
        ''' Get a card from fab API.
        :params id: The identifier of the card.
        :return: Returns a dictionary with the card's data.
        :rtype: dict
        '''
        return self.get(path=f'/cards/{id}')

    def decks(self, slug: str = 'NmzrmMWV') -> dict:
        return self.get(path=f'/decks/{slug}')


pyfabdb/client.py
import json

import requests

from pyfabdb.util import format_url


class Client:
    """
    The client class for communicating with an API.
    """

    def __init__(self, url: str = 'https://api.fabdb.net', header: dict = {'Accept': 'application/json'}):
        self.url = url
        self.header = header
        self.session = requests.Session()

        if self.header is None or not isinstance(self.header, dict):
            self.__set_default_header()

    def __set_default_header(self) -> None:
        self.session.headers.update({'Accept': 'application/json'})

    def _request(self, method: str, base_url: str, path: str, data: dict = {}, header: dict = {}, params: dict = {}):
        if path is not None and base_url is not None:
            url = format_url(self.url, path)
        elif path is not None and base_url is None:
            url = format_url(self.url, path)
        elif path is None:
            raise ValueError('The argument `path` cannot be None.')

        if data is not None:
            data = json.dumps(data)

        if header is not None:
            if not isinstance(header, dict):
                header = json.loads(header)
            self.session.headers.update(header)
        else:
            self.__set_default_header()

        response = self.session.request(method=method, url=url, data=data, headers=self.session.headers, params=params)

        try:
            response.raise_for_status()
        except requests.exceptions.HTTPError as e:
            raise(e)
        except requests.exceptions.RequestException as e:
            raise(e)
        except requests.exceptions.ConnectionError as e:
            raise(e)
        except requests.exceptions.Timeout as e:
            raise(e)

        if response.text != '':
            try:
                return response.json()
            except json.decoder.JSONDecodeError:
                return response.text
        else:
            return response.status_code

    def get(self, base_url: str = '', path: str = '', data: dict = {}, header: dict = {}, params: dict = {}):
        return self._request(method='GET', base_url=base_url, path=path, data=data, header=header, params=params)

    def post(self, base_url: str = '', path: str = '', data: dict = {}, header: dict = {}, params: dict = {}):
        return self._request(method='POST', base_url=base_url, path=path, data=data, header=header, params=params)

    def delete(self, base_url: str = '', path: str = '', data: dict = {}, header: dict = {}, params: dict = {}):
        return self._request(method='DELETE', base_url=base_url, path=path, data=data, header=header, params=params)

    def patch(self, base_url: str = '', path: str = '', data: dict = {}, header: dict = {}, params: dict = {}):
        return self._request(method='PATCH', base_url=base_url, path=path, data=data, header=header, params=params)

    def put(self, base_url: str = '', path: str = '', data: dict = {}, header: dict = {}, params: dict = {}):
        return self._request(method='PUT', base_url=base_url, path=path, data=data, header=header, params=params)


pyfabdb/pyfabdb.py
from pyfabdb.client import Client


class PyFabdb(Client):
    def __init__(self, url: str = 'https://api.fabdb.net', header: dict = {'Accept': 'application/json'}):
        super().__init__(url=url, header=header)

    def cards(self, page: int = 1, per_page: int = 25, keywords: str = '', pitch: str = '', cost: str = '', card_class: str = '', rarity: str = '', set: str = '') -> dict:
        ''' Get cards from fabdb API.
        :param page: When paginating through data sets, specifies the page.
            Integer.
        :param per_page: Specify the number of records per result set.
            Integer. Max-100
        :param keywords: A string representing a search term:
            String
        :params pitch: Search for a card with a given pitch.
            String. Valid values are: (1, 2, or 3)
        :params cost: Search for a card with a given cost.
            String. Valid values are: (1, 2, 3, or 4+)
        :params card_class: Search for a card that matches the required hero class.
           String. Valid values are: brute, guardian, mechanologist, ninja, ranger, runeblade, warrior, wizard.
        :params rarity: Search for a card that matches the specified rarity.
            String. Valid values are: C, R, S, T, L, F, P
        :params set: Search for cards from a given set.
            String. Valid values are: WTR, ARC, CRU, MON
        :return: A list of dictionaries with the cards' data.
        :rtype: dict{}

            Example Returned Data:
            {
                "current_page":19,
                "data": [
                    {
                        "identifier":"WTR219",
                        "name":"Nimblism",
                        "keywords":["generic","action"],
                        "stats":{"cost":"0","defense":"2","resource":"2"},
                    },
                    ...
                ],
                "first_page_url":"/cards?per_page=25&page=1",
                "from":451,
                "next_page_url":null,
                "path":"/cards",
                "per_page":"25",
                "prev_page_url":"/cards?per_page=25&page=18",
                "to":457
            }
        '''

        payload = {}

        if page:
            payload['page'] = page
        if per_page:
            payload['per_page'] = per_page
        if keywords:
            payload['keywords'] = keywords
        if pitch:
            payload['pitch'] = pitch
        if cost:
            payload['cost'] = cost
        if card_class:
            payload['class'] = card_class
        if rarity:
            payload['rarity'] = rarity
        if set:
            payload['set'] = set

        return self.get(path='/cards', params=payload)

    def get_card(self, id: str = 'absorb-in-aether-red') -> dict:
        ''' Get a card from fabdb API.
        :params id: The identifier of the card.
        :return: Returns a dictionary with the card's data.
        :rtype: dict
        '''
        payload = {}
        if id:
            payload['id'] = id

        return self.get(path=f'/cards/{id}', params=payload)

    def decks(self, slug: str = 'eLxddlzb') -> dict:
        ''' Get a deck from fabdb API.
        :params slug: The slug of the deck.
        :return: Returns a dictionary with the deck's data.
        :rtype: dict
        '''
        return self.get(path=f'/decks/{slug}')
